course_UvA-DL/activation-functions/Activation_Functions.py

```diff
@@ -0,0 +1,824 @@
+#
+# The tanh shows a more diverse behavior.
+# While for the input layer we experience a larger amount of neurons to be close to -1 and 1, where the gradients are close to zero, the activations in the two consecutive layers are closer to zero.
+# This is probably because the input layers look for specific features in the input image, and the consecutive layers combine those together.
+# The activations for the last layer are again more biased to the extreme points because the classification layer can be seen as a weighted average of those values (the gradients push the activations to those extremes).
+#
+# The ReLU has a strong peak at 0, as we initially expected.
+# The effect of having no gradients for negative values is that the network does not have a Gaussian-like distribution after the linear layers, but a longer tail towards the positive values.
+# The LeakyReLU shows a very similar behavior while ELU follows again a more Gaussian-like distribution.
+# The Swish activation seems to lie in between, although it is worth noting that Swish uses significantly higher values than other activation functions (up to 20).
+#
+# As all activation functions show slightly different behavior although
+# obtaining similar performance for our simple network, it becomes
+# apparent that the selection of the "optimal" activation function really
+# depends on many factors, and is not the same for all possible networks.
+
+# %% [markdown]
+# ### Finding dead neurons in ReLU networks
+
+# %% [markdown]
+# One known drawback of the ReLU activation is the occurrence of "dead neurons", i.e. neurons with no gradient for any training input.
+# The issue of dead neurons is that as no gradient is provided for the layer, we cannot train the parameters of this neuron in the previous layer to obtain output values besides zero.
+# For dead neurons to happen, the output value of a specific neuron of the linear layer before the ReLU has to be negative for all input images.
+# Considering the large number of neurons we have in a neural network, it is not unlikely for this to happen.
+#
+# To get a better understanding of how much of a problem this is, and when we need to be careful, we will measure how many dead neurons different networks have.
+# For this, we implement a function which runs the network on the whole
+# training set and records whether a neuron is exactly 0 for all data
+# points or not:
+
+
+# %%
+def measure_number_dead_neurons(net):
+
+    # For each neuron, we create a boolean variable initially set to 1. If it has an activation unequals 0 at any time,
+    # we set this variable to 0. After running through the whole training set, only dead neurons will have a 1.
+    neurons_dead = [
+        torch.ones(layer.weight.shape[0], device=device, dtype=torch.bool) for layer in net.layers[:-1]
+        if isinstance(layer, nn.Linear)
+    ]  # Same shapes as hidden size in BaseNetwork
+
+    net.eval()
+    with torch.no_grad():
```

Comment:
I guess all is with no-grad so you can set it as func wrapper
see: https://pytorch.org/docs/stable/generated/torch.no_grad.html